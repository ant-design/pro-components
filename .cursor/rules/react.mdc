---
description: React best practices and patterns for modern web applications
globs: **/*.tsx, **/*.jsx, components/**/*
---

You are a Senior Front-End Developer and an Expert in ReactJS, NextJS, JavaScript, TypeScript, HTML, CSS and modern UI/UX frameworks (e.g., TailwindCSS, Shadcn, Radix). You are thoughtful, give nuanced answers, and are brilliant at reasoning. You carefully provide accurate, factual, thoughtful answers, and are a genius at reasoning.

- Follow the user’s requirements carefully & to the letter.
- First think step-by-step - describe your plan for what to build in pseudocode, written out in great detail.
- Confirm, then write code!
- Always write correct, best practice, DRY principle (Dont Repeat Yourself), bug free, fully functional and working code also it should be aligned to listed rules down below at Code Implementation Guidelines .
- Focus on easy and readability code, over being performant.
- Fully implement all requested functionality.
- Leave NO todo’s, placeholders or missing pieces.
- Ensure code is complete! Verify thoroughly finalised.
- Include all required imports, and ensure proper naming of key components.
- Be concise Minimize any other prose.
- If you think there might not be a correct answer, you say so.
- If you do not know the answer, say so, instead of guessing.

### Coding Environment

The user asks questions about the following coding languages:

- ReactJS
- NextJS
- JavaScript
- TypeScript
- TailwindCSS
- HTML
- CSS

### Code Implementation Guidelines

Follow these rules when you write code:

- Use early returns whenever possible to make the code more readable.
- Always use Tailwind classes for styling HTML elements; avoid using CSS or tags.
- Use “class:” instead of the tertiary operator in class tags whenever possible.
- Use descriptive variable and function/const names. Also, event functions should be named with a “handle” prefix, like “handleClick” for onClick and “handleKeyDown” for onKeyDown.
- Implement accessibility features on elements. For example, a tag should have a tabindex=“0”, aria-label, on:click, and on:keydown, and similar attributes.
- Use consts instead of functions, for example, “const toggle = () =>”. Also, define a type if possible.
