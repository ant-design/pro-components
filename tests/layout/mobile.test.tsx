import { ProLayout } from '@ant-design/pro-components';
import { cleanup, render, waitFor } from '@testing-library/react';
import { act } from 'react';
import defaultProps from './defaultProps';

afterEach(() => {
  cleanup();
});

describe('mobile BasicLayout', () => {
  beforeAll(() => {
    process.env.NODE_ENV = 'TEST';
    process.env.USE_MEDIA = 'xs';

    Object.defineProperty(global.window, 'matchMedia', {
      value: vi.fn((query) => {
        //  (max-width: 575px)
        return {
          media: query,
          matches: query.includes('max-width: 575px'),
          addListener: vi.fn(),
          removeListener: vi.fn(),
        };
      }),
    });
  });

  afterAll(() => {
    process.env.USE_MEDIA = 'md';
    process.env.NODE_ENV = 'dev';
  });

  it('📱 base use', async () => {
    const html = render(
      <ProLayout {...defaultProps} getContainer={false} onCollapse={() => {}}>
        welcome
      </ProLayout>,
    );

    await waitFor(async () => {
      await html.findAllByText('welcome');
    });
    expect(html.asFragment()).toMatchSnapshot();
  });

  it('📱 collapsed=false', async () => {
    const html = render(
      <ProLayout {...defaultProps} getContainer={false} collapsed={false} />,
    );
    await waitFor(async () => {
      await html.findAllByText('welcome');
    });
    expect(html.asFragment()).toMatchSnapshot();
  });

  it('📱 layout=mix', async () => {
    const html = render(
      <ProLayout
        {...defaultProps}
        getContainer={false}
        layout="mix"
        collapsed={false}
      />,
    );
    await waitFor(async () => {
      await html.findAllByText('welcome');
    });
    expect(html.asFragment()).toMatchSnapshot();
  });

  it('📱 layout=mix and splitMenus', async () => {
    const html = render(
      <ProLayout
        {...defaultProps}
        splitMenus
        getContainer={false}
        layout="mix"
        collapsed={false}
      />,
    );
    await waitFor(async () => {
      await html.findAllByText('welcome');
    });
    expect(html.asFragment()).toMatchSnapshot();
  });

  it('📱 layout menuHeaderRender=false', async () => {
    const html = render(
      <ProLayout
        {...defaultProps}
        collapsed
        getContainer={false}
        layout="mix"
        menuHeaderRender={false}
      >
        welcome
      </ProLayout>,
    );
    await waitFor(async () => {
      await html.findAllByText('welcome');
    });
    expect(html.asFragment()).toMatchSnapshot();
  });

  it('📱 layout menuHeaderRender', async () => {
    const html = render(
      <ProLayout
        {...defaultProps}
        collapsed
        getContainer={false}
        layout="mix"
        menuHeaderRender={() => 'title'}
      >
        welcome
      </ProLayout>,
    );
    await waitFor(async () => {
      await html.findAllByText('welcome');
    });
    expect(html.asFragment()).toMatchSnapshot();
  });

  it('📱 layout menuHeaderRender', async () => {
    const html = render(
      <ProLayout
        {...defaultProps}
        collapsed
        getContainer={false}
        layout="mix"
        menuHeaderRender={() => 'title'}
      >
        welcome
      </ProLayout>,
    );
    await waitFor(async () => {
      await html.findAllByText('welcome');
    });
    expect(html.asFragment()).toMatchSnapshot();
  });

  it('📱 layout collapsedButtonRender', async () => {
    const onCollapse = vi.fn();
    const html = render(
      <ProLayout
        {...defaultProps}
        onCollapse={onCollapse}
        collapsed={false}
        collapsedButtonRender={() => {
          return 'div';
        }}
        getContainer={false}
        layout="mix"
      />,
    );

    await waitFor(async () => {
      await html.findAllByText('div');
    });
    act(() => {
      html.baseElement
        ?.querySelector<HTMLSpanElement>(
          'span.ant-pro-global-header-collapsed-button',
        )
        ?.click();
    });

    await waitFor(async () => {
      await html.findAllByText('welcome');
    });

    act(() => {
      html.baseElement
        ?.querySelector<HTMLDivElement>('div.ant-drawer-mask')
        ?.click();
    });

    await waitFor(async () => {
      await html.findAllByText('welcome');
    });

    waitFor(() => {
      expect(onCollapse).toHaveBeenCalled();
    });
  });
});
