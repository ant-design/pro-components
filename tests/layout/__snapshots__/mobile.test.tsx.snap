// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mobile BasicLayout > ğŸ“± collapsed=false 1`] = `<DocumentFragment />`;

exports[`mobile BasicLayout > ğŸ“± layout menuHeaderRender 1`] = `<DocumentFragment />`;

exports[`mobile BasicLayout > ğŸ“± layout menuHeaderRender 2`] = `<DocumentFragment />`;

exports[`mobile BasicLayout > ğŸ“± layout menuHeaderRender 3`] = `<DocumentFragment />`;

exports[`mobile BasicLayout > ğŸ“± layout menuHeaderRender 4`] = `<DocumentFragment />`;

exports[`mobile BasicLayout > ğŸ“± layout menuHeaderRender 5`] = `
<DocumentFragment>
  <div
    class="ant-design-pro ant-pro-layout screen-xs ant-pro-layout-fix-siderbar ant-pro-layout-mix"
  >
    <div
      class="ant-pro-layout-bg-list"
    />
    <div
      class="ant-layout"
      style="min-height: 100%; flex-direction: row;"
    >
      <div
        class="ant-pro-layout-container"
        style="position: relative;"
      >
        <header
          class="ant-layout-header"
          style="height: 56px; line-height: 56px; background-color: transparent; z-index: 19;"
        />
        <header
          class="ant-layout-header ant-pro-layout-header ant-pro-layout-header-fixed-header ant-pro-layout-header-mix ant-pro-layout-header-header"
        >
          <div
            class="ant-pro-global-header"
          >
            <span
              class="ant-pro-global-header-collapsed-button"
            >
              <span
                aria-label="menu"
                class="anticon anticon-menu"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="menu"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M904 160H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8zm0 624H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8zm0-312H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
            </span>
            title
            <div
              style="flex-grow: 1; flex-shrink: 1; flex-basis: 0%;"
            />
          </div>
        </header>
        <main
          class="ant-layout-content ant-pro-layout-content ant-pro-layout-has-header"
        >
          welcome
        </main>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`mobile BasicLayout > ğŸ“± layout menuHeaderRender 6`] = `
<DocumentFragment>
  <div
    class="ant-design-pro ant-pro-layout screen-xs ant-pro-layout-fix-siderbar ant-pro-layout-mix"
  >
    <div
      class="ant-pro-layout-bg-list"
    />
    <div
      class="ant-layout"
      style="min-height: 100%; flex-direction: row;"
    >
      <div
        class="ant-pro-layout-container"
        style="position: relative;"
      >
        <header
          class="ant-layout-header"
          style="height: 56px; line-height: 56px; background-color: transparent; z-index: 19;"
        />
        <header
          class="ant-layout-header ant-pro-layout-header ant-pro-layout-header-fixed-header ant-pro-layout-header-mix ant-pro-layout-header-header"
        >
          <div
            class="ant-pro-global-header"
          >
            <span
              class="ant-pro-global-header-collapsed-button"
            >
              <span
                aria-label="menu"
                class="anticon anticon-menu"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="menu"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M904 160H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8zm0 624H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8zm0-312H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
            </span>
            title
            <div
              style="flex-grow: 1; flex-shrink: 1; flex-basis: 0%;"
            />
          </div>
        </header>
        <main
          class="ant-layout-content ant-pro-layout-content ant-pro-layout-has-header"
        >
          welcome
        </main>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
